src/wallet/wallet.cpp

```diff
@@ -1457,12 +1511,50 @@ mapSproutNoteData_t CWallet::FindMyNotes(const CTransaction& tx) const
     return noteData;
 }
 
+
+/**
+ * Finds all output notes in the given transaction that have been sent to
+ * SaplingPaymentAddresses in this wallet.
+ *
+ * It should never be necessary to call this method with a CWalletTx, because
+ * the result of FindMySaplingNotes (for the addresses available at the time) will
+ * already have been cached in CWalletTx.mapSaplingNoteData.
+ */
+mapSaplingNoteData_t CWallet::FindMySaplingNotes(const CTransaction &tx) const
+{
+    LOCK(cs_SpendingKeyStore);
+    uint256 hash = tx.GetHash();
+
+    mapSaplingNoteData_t noteData;
+
+    // Protocol Spec: 4.19 Block Chain Scanning (Sapling)
+    for (uint32_t i = 0; i < tx.vShieldedOutput.size(); ++i) {
+        const OutputDescription output = tx.vShieldedOutput[i];
+        for (auto it = mapSaplingIncomingViewingKeys.begin(); it != mapSaplingIncomingViewingKeys.end(); ++it) {
+            SaplingIncomingViewingKey ivk = it->second;
+            auto result = SaplingNotePlaintext::decrypt(output.encCiphertext, ivk, output.ephemeralKey);
+            if (!result) {
+                continue;
+            }
+            // We don't cache the nullifier here as computing it requires knowledge of the note position
+            // in the commitment tree, which can only be determined when the transaction has been mined.
+            SaplingOutPoint op {hash, i};
+            SaplingNoteData nd;
+            nd.ivk = ivk;
+            noteData.insert(std::make_pair(op, nd));
+            break;
+        }
+    }
+
+    return noteData;
+}
+
 bool CWallet::IsFromMe(const uint256& nullifier) const
```

Comment:
This should be renamed to ``IsSproutNullifierFromMe`` or something like that.

Reply:
Also, ``CWallet::IsFromMe(const CTransaction& tx)`` needs to be updated to call a corresponding ``IsSaplingNullifierFromMe`` method.

Incidentally, the locking here looks quite inefficient; ideally we wouldn't be locking/unlocking the wallet once for each nullifier in a transaction in ``CWallet::IsFromMe(const CTransaction& tx)``.